edx/edx-enterprise
# ENT-742 Enrollment API

**Description:** 
Introduces a new endpoint for enrolling a single user in a single course run for a given enterprise.

See https://github.com/edx/edx-enterprise/pull/237 for the initial prototype this is based on.

**JIRA:** https://openedx.atlassian.net/browse/ENT-742

**Dependencies:** dependencies on other outstanding PRs, issues, etc. 

**Merge deadline:** 

**Installation instructions:** 

**Testing instructions:**
The endpoint can be accessed using POST calls only at: {base_url}/enterprise/api/v1/enterprise-customer/{enterprise_uuid}/course_enrollments
Only users that have create access to Enterprise Customer models can hit this endpoint. This is temporary until the Authorization story is handled.

The fields to be sent in the POST data are as follows:
lms_user_id - the edX user id
tpa_user_id - the IdP user id
user_email - the user's email address
course_run_id - the course run ID to enroll the user in
course_mode - the course mode to enroll the user in (audit, verified, professional)
email_students - whether or not to email the students to notify them of the enrollment. defaults to false

The endpoint accepts a list of payloads with the above params. So for an individual enrollment call, an example payload would be:
[{
	"course_run_id":"course-v1:edX+DemoX+Demo_Course",
	"course_mode":"verified",
	"user_email":"bexline+enapitest4@edx.org",
	"email_students": true
}]

Multiple would be sent like so:
[
  {
	"course_run_id":"course-v1:edX+DemoX+Demo_Course",
	"course_mode":"verified",
	"user_email":"bexline+enapitest4@edx.org",
	"email_students": true
  },
  {
	"course_run_id":"course-v1:edX+DemoX+Demo_Course",
	"course_mode":"audit",
	"tpa_user_id":"abcdefg"
  }
]

The corresponding response will return an array with dictionaries mapping to each item in the original request. Successful responses will appear as {'detail': 'success'}, and unsuccessful responses will have a message indicating what the failure was for that payload.

Testing basic field validation - course_run_id and course_mode are always required, and at least one of lms_user_id, tpa_user_id, and user_email must be sent. The course_run_id must be part of the Enterprise Customer's Catalog. Making calls that do not meet these requirements should return responses that detail what is missing from the request data.

Note that for calls with multiple payloads, validation is processed serially. If one fails, all other payloads after it in the sequence will not be processed and return an empty value in the response.

Testing user identifier field validation - because the user identifier fields require at least one be specified, and multiple user identifier fields can be sent in one request, ie sending both lms_user_id and tpa_user_id together, there is some additional complexity in the validation. If more than one is specified, the field which results in a successful lookup of an EnterpriseCustomerUser is used, with preference given to lms_user_id, then tpa_user_id, then user_email. If user_email is given and no user is associated with it, and the other fields are not given or are also not associated with a user, a Pending Enterprise User and Pending Enrollment will be created. This fall back does not exist for the other two fields - if they are specified, no user can be found, and user_email is not specified, then the request will return an error indicating this.

Additionally, if the user is enrolled in a paid mode (verified or professional) and the request is attempting to enroll in audit, there will be a message specific to that and no change to enrollment will be performed. This will not impact later payloads if there are multiple sent. The reverse of this is allowed and will affect the enrollment, meaning a user already enrolled in audit or honor can be upgraded to verified/professional through this endpoint. Enrollment calls that attempt to enroll the user in their same mode will behave as if successful.

Testing enrolled user experience - this should behave as it does for users enrolled through the django admin interface. When they log in, they will see the course on their dashboard. If data sharing consent is enabled, they will be prompted for consent when they try to access the course for the first time. For pending users, this will happen after they register with the email they were enrolled with.

Testing email_students flag - if this is set to True, it will behave as the django admin interface and send an email using the same template to new or existing users about their enrollment.

Example Calls and their expected behaviors:
using a user that does not have add EnterpriseCustomer model access - Should give a message about insufficient permission.
no parameters - Should give a message about course_run_id and course_mode being required.
call with course_mode and course_run_id only - Should give a message about specifying one of the user identifier fields.
call with course_mode, course_run_id, and lms_user_id and/or tpa_user_id without an existing user - Should give a message indicating that no user could be found and user_email not being present.
call with course_mode, course_run_id, and user_email without an existing user - Should return a successful response, with a corresponding new PendingEnterpriseCustomerUser and PendingEnrollment created.
call with course_mode, course_run_id, and lms_user_id/tpa_user_id/user_mail with an existing user - Should return a successful response, and a new EnterpriseCourseEnrollment with the corresponding user and course should be created.

Remaining items:
*Links from the email do not trigger SSO for the Enterprise Customer. Depends on work from https://openedx.atlassian.net/browse/ENT-729. Once that is completed, we can change the url sent in emails to learners to include a tpa_hint. This might land separately from this PR depending on timing.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.
